Bluetooth: Host: Reassemble extended advertising reports

hermabe opened the pull request:
This is a simplified version of #39352, which does not support interleaving of fragmented advertising reports from multiple advertisers. Because of this we do not need to keep a list of all seen advertisers and some simplifications can be done.

Advertising reports from legacy PDUs or complete extended advertising reports may be interleaved as these do not require reassembly.

hermabe committed:
Bluetooth: Host: Move ext adv property conversion to separate function

How the conversion worked was unclear. Moving the conversion out to a
separate function and documenting how it works makes the code easier to
understand.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe commented on tests/bluetooth/host_long_adv_recv/src/main.c:
```diff
@@ -0,0 +1,476 @@
+	send_adv_report(&report_b_2); // Interleaved short extended adv report
+	send_adv_report(&report_a_2);
+
+	// Check that fragmented adv reports from one peer are received,
+	// and that interleaved fragmented adv reports from other peers are discarded
+	ztest_returns_value(get_expected_data, &report_a_combined.data[0]);
+	ztest_returns_value(get_expected_length, report_a_combined.length); // Expect only a
+	ztest_returns_value(get_expected_data, &report_b_2.data[0]);
+	ztest_returns_value(get_expected_length, report_b_2.length); // Expect b, INCOMPLETE REPORT
+	send_adv_report(&report_a_1);
+	send_adv_report(&report_b_1); // Interleaved fragmented adv report, NOT SUPPORTED
+	send_adv_report(&report_a_2);
+	send_adv_report(&report_b_2);
```
Should I keep this to document the current behavior or remove it as interleaving is not currently supported? If interleaving was supported, `report_b_combined` would be expected instead of `report_b_2`.

hermabe commented on subsys/bluetooth/host/scan.c:
```diff
@@ -507,50 +551,154 @@ static uint8_t get_adv_type(uint8_t evt_type)
 	}
 }
 
+/* Convert extended adv report evt_type field to adv props */
+static uint16_t get_adv_props_extended(uint16_t evt_type)
+{
+	/* Converts from BT_HCI_LE_ADV_EVT_TYPE_* to BT_GAP_ADV_PROP_*
+	 * The first 4 bits are the same (conn, scan, direct, scan_rsp).
+	 * Bit 4 must be flipped as the meaning of 1 is opposite (legacy -> extended)
+	 * The rest of the bits are zeroed out.
+	 */
+	return (evt_type ^ BT_HCI_LE_ADV_EVT_TYPE_LEGACY) & BIT_MASK(5);
+}
```
I moved this check to a separate function with a comment to make the conversion clearer. The logic from before seems correct, but was a bit hard to decode. @alwa-nordic Had some questions about this.

hermabe commented on subsys/bluetooth/host/scan.c:
```diff
@@ -361,7 +405,7 @@ static void check_pending_conn(const bt_addr_le_t *id_addr,
 
 /* Convert Legacy adv report evt_type field to adv props */
-static uint8_t get_adv_props(uint8_t evt_type)
+static uint8_t get_adv_props_legacy(uint8_t evt_type)
```
Renamed to not confuse with the new function.

hermabe committed:
Bluetooth: Host: Reassemble extended advertising reports

The host reassembles fragmented advertising reports from the controller.

Non-complete advertising reports from different advertisers may not be
interleaved. If non-complete advertising reports from an advertiser
is received while advertising reports from another advertiser is
reassembled, an error message is logged and the advertising report is
discarded. Future scan results may be incomplete.

Advertising reports from legacy PDUs or complete extended advertising
reports may be interleaved as these do not require reassembly.

If the controller sends more advertising data than fits in the
reassembly buffer, the data is truncated. Further advertising reports
from the advertiser are discarded until the final complete advertising
report is received and discarded.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

carlescufi commented:
@hermabe can you please rebase on top of latest main?

carlescufi commented:
@bewi-nordic @rugeGerritsen  please take a look

carlescufi commented:
This is for Ext Adv, previously was periodic sync: #40596.

Thalley commented on tests/bluetooth/host_long_adv_recv/src/main.c:
```diff
@@ -0,0 +1,482 @@
+		     report_b_2.length);
+
+	/* Check that non-interleaved fragmented adv reports work */
+	ztest_returns_value(get_expected_data, &report_a_combined.data[0]);
```
General comment: Why do `&report_a_combined.data[0]` instead of `report_a_combined.data`?

hermabe replied:
I have seen this in other places in the codebase and I think I had some other indexes during development. Your suggestion is less noisy.

hermabe commented on tests/bluetooth/host_long_adv_recv/src/main.c:
```diff
@@ -0,0 +1,482 @@
+	}
+}
+
+/* Test. */
+static void test_host_long_adv_recv(void)
+{
```
A babblesim test may be better. Here I had to copy a bunch of setup to make HCI work. The reason this isn't a babblesim test is that when I started this ZLL did not support fragmented advertising reports and SDC did not support babblesim.

Thalley replied:
Ah, right, for Babblesim you'd need to have both sides first, yes :)

hermabe commented on subsys/bluetooth/host/scan.c:
```diff
@@ -507,50 +551,157 @@ static uint8_t get_adv_type(uint8_t evt_type)
+			continue;
+		}
+
+		const bool is_new_advertiser =
+			!p_reassembling_advertiser ||
+			!fragmented_advertisers_equal(p_reassembling_advertiser, &evt->addr,
+						      evt->sid);
```
I am personally in favor of declaring things as late as possible, but it seems the rest of the codebase declares everything on top. Simplest change here is to move this one up.

Thalley replied:
There isn't a specific rule for that, and since there isn't then the rule is to follow the file's code style (which seems to be declarations at top) :)

hermabe commented on subsys/bluetooth/host/scan.c:
```diff
@@ -27,6 +28,41 @@ static bt_le_scan_cb_t *scan_dev_found_cb;
+	struct net_buf *reassembly_buffer;
+};
+
+static struct fragmented_advertiser reassembling_advertiser;
+static struct fragmented_advertiser *p_reassembling_advertiser;
```
I am mostly using it as a flag to see if we are currently reassembling anything. A bool flag in the struct may be better.

hermabe replied:
I am thinking if we add an enum for state to the `fragmented_advertiser` struct we can get rid of this pointer and the `reassembly_buffer` pointer. Something like:
```c
enum {
  INACTIVE,
  REASSEMBLING,
  DISCARDING,
} state;
```

Currently `p_reassembling_advertiser` is set to `NULL` to signal that the advertiser is inactive and to `&reassembling_advertiser` to signal that it is reassembling and the `reassembly_buffer` pointer is set to `NULL`  to signal that the advertiser is still active, but the data should be discarded.

Thalley replied:
@hermabe Could we use the buf->len == 0 to determine if a reassembly is in progress?

hermabe replied:
That should work. Still need to flag if the reports should be discarded.

Thalley replied:
Yes :) The change that I suggest here is that instead of checking `p_reassembling_advertiser == NULL` we check `reassembling_advertiser.reassembly_buffer->len == 0` (possibly as a macro).

hermabe replied:
When `reassembling_advertiser.reassembly_buffer` is set to `NULL` to signal that reports should be discarded this does not work. Created the enum mentioned above instead.

hermabe commented on subsys/bluetooth/host/scan.c:
```diff
@@ -507,50 +551,157 @@ static uint8_t get_adv_type(uint8_t evt_type)
+		create_ext_adv_info(evt, &scan_info);
+		if (truncate || ctrl_truncated) {
+			scan_info.adv_props |= BT_GAP_ADV_PROP_REPORT_TRUNCATED;
+		}
```
Truncated data is no longer sent to the upper layers.

hermabe commented on include/bluetooth/gap.h:
```diff
@@ -126,6 +126,12 @@ enum {
+	 * Either the controller or the host was not able to receive or store
+	 * the complete advertisement.
+	 */
+	BT_GAP_ADV_PROP_REPORT_TRUNCATED      = BIT(5),
```
Removed.

Thalley reviewed (approved):
LGTM. The implementation is simple in my opinion which is of course what you went for. The code is readable, and doesn't do more than what we want. Good job :)

hermabe committed:
Bluetooth: Host: Reset scan state on scan stop

bt_le_stop() previously did the same restting as bt_scan_reset(). After
the recent changes they were out of sync.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

alwa-nordic reviewed (approved):
Looks great!

hermabe commented:
@Thalley Are you OK with the last commit I added after you approved so this can be merged?

Thalley commented:
> @Thalley Are you OK with the last commit I added after you approved so this can be merged?

Yes :) Next time you can also re-request a review from reviewers if you feel like the PR needs it.

LingaoM commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -164,8 +164,7 @@ static inline void get_evt_hdr(void)
 		if (rx.evt.evt == BT_HCI_EVT_LE_META_EVENT &&
-		    (rx.hdr[sizeof(*hdr)] == BT_HCI_EVT_LE_ADVERTISING_REPORT ||
-		     rx.hdr[sizeof(*hdr)] == BT_HCI_EVT_LE_EXT_ADVERTISING_REPORT)) {
+		    (rx.hdr[sizeof(*hdr)] == BT_HCI_EVT_LE_ADVERTISING_REPORT)) {
```
Remove `BT_HCI_EVT_LE_EXT_ADVERTISING_REPORT` will cause deadlock in `bt hci driver task`.
